File: src/App/Placement.cpp
/***************************************************************************
 *   Copyright (c) 2002 Jürgen Riegel <juergen.riegel@web.de>              *
 *                                                                         *
 *   This file is part of the FreeCAD CAx development system.              *
 *                                                                         *
 *   This library is free software; you can redistribute it and/or         *
 *   modify it under the terms of the GNU Library General Public           *
 *   License as published by the Free Software Foundation; either          *
 *   version 2 of the License, or (at your option) any later version.      *
 *                                                                         *
 *   This library  is distributed in the hope that it will be useful,      *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU Library General Public License for more details.                  *
 *                                                                         *
 *   You should have received a copy of the GNU Library General Public     *
 *   License along with this library; see the file COPYING.LIB. If not,    *
 *   write to the Free Software Foundation, Inc., 59 Temple Place,         *
 *   Suite 330, Boston, MA  02111-1307, USA                                *
 *                                                                         *
 ***************************************************************************/


#include "PreCompiled.h"

#include <Base/Placement.h>

#include "Placement.h"

using namespace App;


PROPERTY_SOURCE(App::Placement, App::GeoFeature)


//===========================================================================
// Feature
//===========================================================================

Placement::Placement(void)
{

}

Placement::~Placement(void)
{
}



// Python feature ---------------------------------------------------------
namespace App {
PROPERTY_SOURCE_TEMPLATE(App::PlacementPython, App::Placement)
template<> const char* App::PlacementPython::getViewProviderName(void) const {
   return "Gui::ViewProviderPlacementPython";
}
template class AppExport FeaturePythonT<App::Placement>;
}

File: src/Mod/Forms/Forms/placement.py
# SPDX-License-Identifier: LGPL-2.1-or-later
# /**************************************************************************
#                                                                           *
#    Copyright (c) 2026 AstoCAD     <hello@astocad.com>                     *
#                                                                           *
#    This file is part of FreeCAD.                                          *
#                                                                           *
#    FreeCAD is free software: you can redistribute it and/or modify it     *
#    under the terms of the GNU Lesser General Public License as            *
#    published by the Free Software Foundation, either version 2.1 of the   *
#    License, or (at your option) any later version.                        *
#                                                                           *
#    FreeCAD is distributed in the hope that it will be useful, but         *
#    WITHOUT ANY WARRANTY; without even the implied warranty of             *
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU       *
#    Lesser General Public License for more details.                        *
#                                                                           *
#    You should have received a copy of the GNU Lesser General Public       *
#    License along with FreeCAD. If not, see                                *
#    <https://www.gnu.org/licenses/>.                                       *
#                                                                           *
# **************************************************************************/

"""Placement helpers for Forms objects and linked reference geometry."""


def global_placement(obj):
    """Return *obj*'s placement in its unique document-tree context.

    ``GeoFeature.getGlobalPlacement`` does not account for Links and is being
    removed from FreeCAD. ``getGlobalPlacementOf`` needs the selection root
    and path explicitly, which are available through ``Parents`` for the
    normal single-parent Forms workflows.
    """
    parents = getattr(obj, "Parents", ())
    if len(parents) == 1 and len(parents[0]) == 2:
        root, subname = parents[0]
    else:
        root, subname = obj, ""
    try:
        return obj.getGlobalPlacementOf(obj, root, str(subname))
    except (AttributeError, RuntimeError, TypeError):
        # Keep documents usable on FreeCAD versions predating the replacement
        # API. This fallback can be removed with FreeCAD's deprecated method.
        try:
            return obj.getGlobalPlacement()
        except (AttributeError, RuntimeError):
            return obj.Placement


__all__ = ["global_placement"]
